### PR TITLE
Forward MHA find2.0 interface and implementation

### PR DESCRIPTION
**Pull request by CAHEK7**
Dropout is missing right now, because it requires random state initialization. It will be added with the backward pass.

Interface and numeric tests will be added in separated PRs.

### CODE REVIEW

**Comment by CAHEK7** [comment]
I did some preliminary local testing (based on https://github.com/ROCm/MIOpen/pull/2563).
Further test integration will be done in separated PR.

**Comment by CAHEK7** [comment]
@atamazov @DrizztDoUrden @amberhassaan can you take a look?

File: include/miopen/miopen.h
Comment (atamazov) [inline include/miopen/miopen.h]: [Recommendation] This is still BETA API, isn't it?

File: include/miopen/miopen.h
Comment (atamazov) [inline include/miopen/miopen.h]: Ditto

File: include/miopen/miopen.h
Comment (atamazov) [inline include/miopen/miopen.h]: [Recommendation] Here and everywhere: Despite the fact that the old APIs are done differently, I prefer keeping camelCase even for acronyms:
```suggestion
MIOPEN_DECLARE_OBJECT(miopenMhaDescriptor);
```
What do you think?

File: include/miopen/miopen.h
Comment (atamazov) [inline include/miopen/miopen.h]: Formatting.

File: src/CMakeLists.txt
Comment (atamazov) [inline src/CMakeLists.txt]: Sorting
```suggestion
    lrn_api.cpp
    mha/mha_descriptor.cpp
    mha/problem_description.cpp
```

File: src/CMakeLists.txt
Comment (atamazov) [inline src/CMakeLists.txt]: ditto
```suggestion
        kernels/miopen_rocrand.hpp
        kernels/miopen_type_traits.hpp
        kernels/miopen_utility.hpp
```

File: src/include/miopen/mha/invoke_params.hpp
Comment (atamazov) [inline src/include/miopen/mha/invoke_params.hpp]: Can you please remind me why do we need a copy of `MHAInputDescsForward`? It's size is not that small.

/cc @DrizztDoUrden

Comment (atamazov) [inline src/include/miopen/mha/invoke_params.hpp]: [Recommendation] I suspect that all data members can be made const.

File: src/mha/mha_descriptor.cpp
Comment (atamazov) [inline src/mha/mha_descriptor.cpp:26]: Please add license header.

File: src/include/miopen/mha/mha_descriptor.hpp
Comment (atamazov) [inline src/include/miopen/mha/mha_descriptor.hpp]: Fixed ordering, path to mha.hpp, see #2638 
```suggestion
#include <miopen/common.hpp>
#include <miopen/mha/mha.hpp>
#include <miopen/miopen.h>
#include <miopen/object.hpp>

#include <nlohmann/json_fwd.hpp>
```

**Review by atamazov — CHANGES_REQUESTED** [review]
shallowish partial review

File: include/miopen/miopen.h
Comment (Vsevolod1983) [reply include/miopen/miopen.h]: Actually, its not. But we can discuss it if there are different opinions

File: src/include/miopen/mha/invoke_params.hpp
Comment (Vsevolod1983) [reply src/include/miopen/mha/invoke_params.hpp]: It's better to not rely on an user of this class. I think we cannot always be sure that such a  reference is valid.

Comment (CAHEK7) [reply src/include/miopen/mha/invoke_params.hpp]: I agree, but...
It can be vulnerable to `std::laundry`, especially in C++20 and if there is any kind of type-erasure or reinterpretation. It's a gray zone, I did not dig deep enough.

Comment (Vsevolod1983) [reply src/include/miopen/mha/invoke_params.hpp]: I made it const actually, like in other invoke params

File: include/miopen/miopen.h
Comment (atamazov) [reply include/miopen/miopen.h]: [resolved] @Vsevolod1983 pls. hide this comment

File: include/miopen/miopen.h
Comment (atamazov) [reply include/miopen/miopen.h]: [resolved] @Vsevolod1983 pls. hide this comment

File: include/miopen/miopen.h
Comment (atamazov) [reply include/miopen/miopen.h]: [resolved] @Vsevolod1983 pls. hide this comment

File: include/miopen/miopen.h
Comment (atamazov) [reply include/miopen/miopen.h]: [resolved] @Vsevolod1983 pls. hide this comment

File: src/CMakeLists.txt
Comment (atamazov) [reply src/CMakeLists.txt]: [resolved] @Vsevolod1983 pls. hide this comment

File: src/CMakeLists.txt
Comment (atamazov) [reply src/CMakeLists.txt]: [resolved] @Vsevolod1983 pls. hide this comment

File: src/include/miopen/mha/invoke_params.hpp
Comment (atamazov) [reply src/include/miopen/mha/invoke_params.hpp]: @Vsevolod1983 @DrizztDoUrden I do not see why. Let's discuss this topic offline.

File: src/mha/mha_descriptor.cpp
Comment (Vsevolod1983) [reply src/mha/mha_descriptor.cpp:26]: Added

File: src/include/miopen/mha/invoke_params.hpp
Comment (atamazov) [reply src/include/miopen/mha/invoke_params.hpp]: [resolved] @Vsevolod1983 pls. hide this comment

File: src/mha/mha_descriptor.cpp
Comment (atamazov) [reply src/mha/mha_descriptor.cpp:26]: [resolved] @Vsevolod1983 pls. hide this comment

File: src/include/miopen/mha/mha_descriptor.hpp
Comment (atamazov) [reply src/include/miopen/mha/mha_descriptor.hpp]: [resolved] @Vsevolod1983 pls. hide this comment

File: src/include/miopen/mha/invoke_params.hpp
Comment (Vsevolod1983) [reply src/include/miopen/mha/invoke_params.hpp]: @atamazov 
Actually, pretty same reason why we don't use const references to Tensor Descriptors but use copy of Tensor Descriptors instead.
Was discussed here:
https://github.com/ROCm/MIOpen/pull/2671#discussion_r1466920140

Comment (Vsevolod1983) [reply src/include/miopen/mha/invoke_params.hpp]: Could we merge this PR to unblock other PRs creation ?

Comment (atamazov) [reply src/include/miopen/mha/invoke_params.hpp]: @Vsevolod1983 This comment cannot be considered resolved, but this is not a blocker to merging this PR.

Comment (atamazov) [reply src/include/miopen/mha/invoke_params.hpp]: [future][performance] To be discussed offline.

File: src/kernels/miopen_rocrand.hpp
Comment (atamazov) [inline src/kernels/miopen_rocrand.hpp:29]: [Q] Do we plan to re-use rocrand?

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [inline src/kernels/MIOpenSoftmaxAttn.cpp:40]: [informative] What about `std::max()`?

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [inline src/kernels/MIOpenSoftmaxAttn.cpp:45]: [R] Typo etc
```suggestion
/// Atomically calculates maximum of non-negative ordered values.
/// Produces wrong results for negatve values or nans,
/// but it is a final amax reducton step and we expect only non-negative ordered values.
```

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [inline src/kernels/MIOpenSoftmaxAttn.cpp:275]: [R]
```suggestion
        // It is supposed to be maximum of absolute values,
        // however we do not need abs() because expf() above produces
        // non-negative value. Plain max() is enough.
```

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [inline src/kernels/MIOpenSoftmaxAttn.cpp:365]: [R] Ditto

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [inline src/kernels/MIOpenSoftmaxAttn.cpp:46]: [clarity]
```suggestion
__device__ float atomicMaxOfNonNegative(float* addr, float value)
```

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [inline src/kernels/MIOpenSoftmaxAttn.cpp:48]: [R] For IEEE-754-compatible formats, when two floats are compared as UNSIGNED, the following is valid:
```
+0.0 < +denorm < +norm < +INF < +SNAN < +QNAN < -0.0 < -denorm < -norm < -INF < -SNAN < -QNAN
```
So if you use unsigned comparison, then both positive and negative NaNs and INFs are considered greater than positive normal and denormal numbers, which is good for MAX operation. But -0.0 is > than +norm/+denorm, which is wrong.

When two floats are compared as SIGNED:
```
-0.0 < -denorm < -norm < -INF < -SNAN < -QNAN < +0.0 < +denorm < +norm < +INF < +SNAN < +QNAN < 
```
This produces "valid" results with -0.0, but results with -NANs/-INFs will be incorrect (which means that -NANs/-INFs may disappear after atomic max, and that may camouflage correctness issues, if any).

Clearing of MSB guarantees valid result with both -0.0 and -NAN/-INF, but this is impossible to do during atomic max. Therefore I recommend careful analyzing of possible inputs:
- If we are pretty sure that -0.0/-INF/-NAN can't appear on input, then we are fine with both signed and unsigned comparison.
- If there is possibility of -0.0, then I would recommend using SIGNED comparison.
- If only -INF/-NAN may appear, then it is better to use UNSIGNED
- If everything is possible (-0.0/-INF/-NAN) then let's use SIGNED. This guarantees valid results with -0.0, but who really cares about disappearing -INF/-NAN? Seriously, if we have -INF/-NAN then it is likely that we also have +INF/+NAN, which do not disappear.

It's up to you what to do with the above considerations. I guess that at least you may wish to add some comments to the code.

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [inline src/kernels/MIOpenSoftmaxAttn.cpp:437]: [R] github shows missing EOL

**Review by atamazov — COMMENTED** [review]
Device code reviewed. No objections.

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [reply src/kernels/MIOpenSoftmaxAttn.cpp:46]: Current name give false impression that it performs `max(abs(...))`.

File: src/kernels/miopen_rocrand.hpp
Comment (CAHEK7) [reply src/kernels/miopen_rocrand.hpp:29]: We do, at least its kernel code.
RNG node allows the user to run different prng algorithms and we are not going to implement all of them.

Host code does not allow to initialize prng GPU state.

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (CAHEK7) [reply src/kernels/MIOpenSoftmaxAttn.cpp:48]: I'm aware about this behavior. This function must be called only for non-negative-ordered numbers.

It does not work with NANs, since `fmax(val, NAN) = val` (because all the comparisons `val` vs `NAN` if always `false`), but this function returns a `NAN`, and it does not work with the negative values, which has already been stated in the comment.
C++ std: https://en.cppreference.com/w/cpp/numeric/math/fmax
CUDA: https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE_1g6e7516db46be25c33fb26e203287f2a3

As for the input data - prior to that atomic operation we do `fmax` operation with initial `0` and even if there is any `NAN` in the input data, it will be filtered and replaced by 0 or any other positive value.

Comment (atamazov) [reply src/kernels/MIOpenSoftmaxAttn.cpp:48]: @CAHEK7 
> This function must be called only for non-negative-ordered numbers.

Yes, but the question is: can we provide a guarantee that this function will be called with such a numbers?

If not, then another question: Do we want NaNs to disappear or to remain? The answer depend on what is required from the softmax primitive.

> As for the input data - prior to that atomic operation we do `fmax` operation with initial `0` and even if there is any `NAN` in the input data, it will be filtered and replaced by 0 or any other positive value.

Is it what we really want?

File: src/kernels/miopen_rocrand.hpp
Comment (atamazov) [reply src/kernels/miopen_rocrand.hpp:29]: [resolved]

**Review by atamazov — APPROVED** [review]
🟢 Good to merge provided that the following comments will be eventually resolved either in discussion threads right here, or in the future PR(s):
- [ ] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1543138431 - [Q] copying of PD into Invoker
- [x] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1546628019 - [Q] max_op: std:max or...?
- [x] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1546632569 - [R] comments
- [x] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1546643193 - [R] comments
- [x] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1546645308 - [R] comments
- [x] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1546648661 - function name must be fixed
- [x] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1546692620 - [Q] what we want to do with input and intermediate NANs/INFs.
- [ ] https://github.com/ROCm/MIOpen/pull/2819#discussion_r1548102760 - [Bug] Write required workspace size into Solution

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (CAHEK7) [reply src/kernels/MIOpenSoftmaxAttn.cpp:48]: > Yes, but the question is: can we provide a guarantee that this function will be called with such a numbers?

Yes, because the value prior this atomic operation acquired from `fmax(0, val)` which filters all the `NANs`

> Is it what we really want?

Kind of, the only case where it can produce different result against normal `fmax` reduction is the buffer of all `NANs` - we will produce 0. In the case where we have at least 1 normal number we will match.

The documentation does not say anything explicit about that case.

File: src/include/miopen/mha/invoke_params.hpp
Comment (amberhassaan) [inline src/include/miopen/mha/invoke_params.hpp:36]: minor suggestion: rename to `MhaInvokeParams` to be inline with other derived classes of `InvokeParams`.

File: src/solver/mha/mha_solver.cpp
Comment (amberhassaan) [inline src/solver/mha/mha_solver.cpp:300]: not sure if I saw it but `GetSolution()` should also store a size value in `result.workspace_size`.

**Review by amberhassaan — COMMENTED** [review]
Looks ok. Please add gtests before merging.

**Comment by Vsevolod1983** [comment]
> Looks ok. Please add gtests before merging.

GTest for MHA Find 2.0 wrapper is in a separate branch. We need to merge this PR in order to create a GTest PR

File: src/include/miopen/mha/invoke_params.hpp
Comment (Vsevolod1983) [reply src/include/miopen/mha/invoke_params.hpp:36]: Well, I 'm looking at other primitives....
softmax - just "InvokeParams" class under softmax namespace.
pooling  - "FwdInvokeParams" class under pooling namespace
reduce - "InvokeParams" class under reduce namespace,
etc

**Review by amberhassaan — APPROVED** [review]

File: src/include/miopen/mha/invoke_params.hpp
Comment (amberhassaan) [reply src/include/miopen/mha/invoke_params.hpp:36]: I guess then we're not consistent in our repo. We have `WrwInvokeParams` and `DataInvokeParams` for convolution and then fusion stuff is named similarly I believe.

Comment (amberhassaan) [reply src/include/miopen/mha/invoke_params.hpp:36]: I am of the opinion that we should name the classes explicitly instead of relying on what namespace a certain `InvokeParams` is in and sometimes it may also lead to collisions that are not detected. 
CC: @atamazov , @DrizztDoUrden

Comment (CAHEK7) [reply src/include/miopen/mha/invoke_params.hpp:36]: @amberhassaan 
What kind of non-detected collisions do you mean?
If we do not overuse `using namespace` - namespace is a good solution.

**Comment by CAHEK7** [comment]
@junliume if you have a change, could we merge it. It blocks few other PRs with the tests and backward pass as well.

**Review by junliume — APPROVED** [review]

File: src/include/miopen/mha/invoke_params.hpp
Comment (atamazov) [reply src/include/miopen/mha/invoke_params.hpp:36]: @Vsevolod1983 

> softmax - just "InvokeParams" class under softmax namespace.
> pooling - "FwdInvokeParams" class under pooling namespace

Maybe this is because pooling requires different InvokeParams for Bwd, while softmax doesn't?

@amberhassaan @CAHEK7 
> If we do not overuse `using namespace` - namespace is a good solution.

I totally agree.

File: src/solver/mha/mha_solver.cpp
Comment (atamazov) [reply src/solver/mha/mha_solver.cpp:300]: @CAHEK7 @amberhassaan ...except that if the solution does not require workspace, then it may rely on the default (which is 0).

File: src/kernels/MIOpenSoftmaxAttn.cpp
Comment (atamazov) [reply src/kernels/MIOpenSoftmaxAttn.cpp:48]: @CAHEK7 Let me summarize: **_The primitive does not guarantee correct (well-defined) results when input is invalid (i.e. contains INFs or NANs)._** Correct behavior in this case means that if input is invalid then output should be invalid as well. Invalid output means INF or NAN; it is questionable that output 0 should be considered invalid (it is meaningless, but formally valid).

**_The current implementation is such that intermediate INFs and NANs do not appear provided that the input is valid._**

Please unhide this comment.

[resolved]

**Pull request merged**
